File: src/common/py/core/messaging/command_reply.py
import typing
import uuid
from dataclasses import dataclass, field
from enum import IntEnum, auto

from .message import Message, Trace


@dataclass(frozen=True, kw_only=True)
class CommandReply(Message):
    """
    A command reply message.

    Every command needs to receive a reply in the form of a ``CommandReply`` message. The reply contains
    information about its ``success``, as well as a text message which is usually used to describe reasons for
    failures.

    Attributes:
        success: Whether the command succeeded.
        message: Arbitrary text, usually used to describe reasons for failures.
        unique: The unique identifier of its corresponding command.
    """
    class FailType(IntEnum):
        """
        Used when a command failed.
        """
        NONE = 0
        TIMEOUT = auto()
        EXCEPTION = auto()
        UNKNOWN = auto()
    
    success: bool = True
    message: str = ""
    
    unique: Trace = field(default_factory=uuid.uuid4)


CommandReplyType = typing.TypeVar("CommandReplyType", bound=CommandReply)  # pylint: disable=invalid-name

CommandDoneCallback = typing.Callable[[CommandReplyType, bool, str], None]
CommandFailCallback = typing.Callable[[CommandReply.FailType, str], None]

File: src/common/web/core/messaging/CommandReply.ts
// @ts-ignore
import { v4 as uuidv4 } from "uuid";

import { Message, type MessageCategory, type Trace } from "./Message";

/**
 * Used when a command failed.
 */
export const enum CommandFailType {
    None = 0,
    Timeout,
    Exception,
    Unknown,
}

/**
 * A command reply message.
 *
 * Every command needs to receive a reply in the form of a ``CommandReply`` message. The reply contains
 * information about its ``success``, as well as a text message which is usually used to describe reasons for
 * failures.
 */
export class CommandReply extends Message {
    public static readonly Category: MessageCategory = "CommandReply";

    public readonly success: boolean = true;
    public readonly message: string = "";

    public readonly unique: Trace = uuidv4();

    /**
     * Gets the global message category.
     */
    public get messageCategory(): MessageCategory {
        return CommandReply.Category;
    }
}

export type CommandDoneCallback = (reply: CommandReply, success: boolean, msg: string) => void;
export type CommandFailCallback = (failType: CommandFailType, msg: string) => void;
